controllers/gitrepository_controller.go

```diff
@@ -87,410 +84,505 @@ func (r *GitRepositoryReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, o
-		return r.reconcileDelete(ctx, repository)
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(obj, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
 	}
 
-	// Return early if the object is suspended.
-	if repository.Spec.Suspend {
-		log.Info("Reconciliation is suspended for this object")
-		return ctrl.Result{}, nil
-	}
+	// Always attempt to patch the object and status after each reconciliation
+	defer func() {
+		// Record the value of the reconciliation request, if any
+		if v, ok := meta.ReconcileAnnotationValue(obj.GetAnnotations()); ok {
+			obj.Status.SetLastHandledReconcileRequest(v)
+		}
 
-	// check dependencies
-	if len(repository.Spec.Include) > 0 {
-		if err := r.checkDependencies(repository); err != nil {
-			repository = sourcev1.GitRepositoryNotReady(repository, meta.DependencyNotReadyReason, err.Error())
-			if err := r.updateStatus(ctx, req, repository.Status); err != nil {
-				log.Error(err, "unable to update status for dependency not ready")
-				return ctrl.Result{Requeue: true}, err
-			}
-			// we can't rely on exponential backoff because it will prolong the execution too much,
-			// instead we requeue on a fix interval.
-			msg := fmt.Sprintf("Dependencies do not meet ready condition, retrying in %s", r.requeueDependency.String())
-			log.Info(msg)
-			r.event(ctx, repository, events.EventSeverityInfo, msg)
-			r.recordReadiness(ctx, repository)
-			return ctrl.Result{RequeueAfter: r.requeueDependency}, nil
+		// Summarize the Ready condition based on abnormalities that may have been observed.
+		conditions.SetSummary(obj,
+			meta.ReadyCondition,
+			conditions.WithConditions(
+				sourcev1.IncludeUnavailableCondition,
+				sourcev1.SourceVerifiedCondition,
+				sourcev1.CheckoutFailedCondition,
+				sourcev1.ArtifactOutdatedCondition,
+				sourcev1.ArtifactUnavailableCondition,
+			),
+			conditions.WithNegativePolarityConditions(
+				sourcev1.ArtifactUnavailableCondition,
+				sourcev1.CheckoutFailedCondition,
+				sourcev1.SourceVerifiedCondition,
+				sourcev1.IncludeUnavailableCondition,
+				sourcev1.ArtifactOutdatedCondition,
+			),
+		)
+
+		// Patch the object, ignoring conflicts on the conditions owned by this controller
+		patchOpts := []patch.Option{
+			patch.WithOwnedConditions{
+				Conditions: []string{
+					sourcev1.ArtifactUnavailableCondition,
+					sourcev1.CheckoutFailedCondition,
+					sourcev1.IncludeUnavailableCondition,
+					sourcev1.ArtifactOutdatedCondition,
+					meta.ReadyCondition,
+					meta.ReconcilingCondition,
+					meta.StalledCondition,
+				},
+			},
 		}
-		log.Info("All dependencies area ready, proceeding with reconciliation")
-	}
 
-	// record reconciliation duration
-	if r.MetricsRecorder != nil {
-		objRef, err := reference.GetReference(r.Scheme, &repository)
-		if err != nil {
-			return ctrl.Result{}, err
+		// Determine if the resource is still being reconciled, or if it has stalled, and record this observation
```

Comment:
This comment confuses me, what does `still being reconciled`, does it mean `not marked for deletion`? Can you please also explain when `result.IsZero()` returns true if `retErr` is nil.

Reply:
We do no longer always return a `RequeueAfter`, as we know that in some cases there is a finite state (e.g. because there is a wrong path configured, or some other type of error that is not solved automatically).

The possible returns are:

1. `Result.RequeueAfter && `err == nil`: we did not run into an issue, and there is a delaye requeue. This means we are no longer actively `Reconciling`, and if the resource is not in a `Ready` state, have stalled.
2. `Result.IsZero() && err == nil`: we did not return a requeue instruction and no error either. This can for example happen when we know an error is persistent, and can only be resolved by a new generation of the resource, the resource should thus be marked as `Stalled`.
3. An error is returned, so we are expecting to try again in a relatively short time frame. In this case the `Reconciling` state can be left untouched.